src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -176,9 +174,12 @@ static bool HasStateFlag(Http2OutputProducer.State state, Http2OutputProducer.St
                         if (actual > 0)
                         {
                             // If we got here it means we're going to cancel the write. Restore any consumed bytes to the connection window.
-                            lock (_windowUpdateLock)
+                            if (actual > int.MaxValue || !TryUpdateConnectionWindow((int)actual))
```

Comment:
What's the significance of the window being greater than int.MaxValue here? Could you add a comment explaining why

Reply:
I was being overly defensive. `actual > int.MaxValue` should never be possible because it would exceed [Http2PeerSettings.MaxWindowSize](https://github.com/dotnet/aspnetcore/blob/3ea008c80d5cc63de7f90ddfd6823b7b006251ff/src/Shared/ServerInfrastructure/Http2/Http2PeerSettings.cs#L10-L14) which is a protocol-defined limit. There's no way Kestrel would ever try to write more than that in one go.

Since I was trying to get rid of extraneous checks elsewhere in this PR, I'll add an assert+comment and do the same here. Having the code make it clear what's actually possible and what's not makes the code easier to reason about later.